File: test/python/test_message_collector.py
#!/usr/bin/env python3
# Copyright 2026 The sim_harness Authors
# SPDX-License-Identifier: Apache-2.0

"""Unit tests for message collector module."""

import pytest

from sim_harness.collector import MessageCollector


# Mock classes for testing without ROS 2
class MockNode:
    """Mock ROS 2 node for testing."""

    def __init__(self):
        self._subscriptions = []

    def create_subscription(self, msg_type, topic, callback, qos_profile):
        """Create a mock subscription."""
        sub = MockSubscription(msg_type, topic, callback)
        self._subscriptions.append(sub)
        return sub

    def destroy_subscription(self, subscription):
        """Destroy a subscription."""
        if subscription in self._subscriptions:
            self._subscriptions.remove(subscription)


class MockSubscription:
    """Mock ROS 2 subscription."""

    def __init__(self, msg_type, topic, callback):
        self.msg_type = msg_type
        self.topic = topic
        self.callback = callback

    def simulate_message(self, msg):
        """Simulate receiving a message."""
        self.callback(msg)


class MockMessage:
    """Mock message for testing."""

    def __init__(self, data):
        self.data = data


class TestMessageCollector:
    """Tests for MessageCollector class."""

    def test_initialization(self):
        """Test collector initialization."""
        node = MockNode()
        collector = MessageCollector(node, "/test_topic", MockMessage)
        assert collector.count() == 0

    def test_collect_messages(self):
        """Test collecting messages."""
        node = MockNode()
        collector = MessageCollector(node, "/test_topic", MockMessage)

        sub = node._subscriptions[0]
        sub.simulate_message(MockMessage("data1"))
        sub.simulate_message(MockMessage("data2"))

        assert collector.count() == 2

        messages = collector.get_messages()
        assert len(messages) == 2
        assert messages[0].data == "data1"
        assert messages[1].data == "data2"

    def test_latest(self):
        """Test getting latest message."""
        node = MockNode()
        collector = MessageCollector(node, "/test_topic", MockMessage)

        assert collector.latest() is None

        sub = node._subscriptions[0]
        sub.simulate_message(MockMessage("first"))
        sub.simulate_message(MockMessage("second"))
        sub.simulate_message(MockMessage("third"))

        latest = collector.latest()
        assert latest is not None
        assert latest.data == "third"

    def test_clear_messages(self):
        """Test clearing collected messages."""
        node = MockNode()
        collector = MessageCollector(node, "/test_topic", MockMessage)

        sub = node._subscriptions[0]
        sub.simulate_message(MockMessage("data"))

        assert collector.count() == 1
        collector.clear()
        assert collector.count() == 0

    def test_max_messages(self):
        """Test max_messages limit."""
        node = MockNode()
        collector = MessageCollector(node, "/test_topic", MockMessage, max_messages=3)

        sub = node._subscriptions[0]
        for i in range(10):
            sub.simulate_message(MockMessage(f"msg_{i}"))

        assert collector.count() == 3

    def test_destroy(self):
        """Test destroying collector subscription."""
        node = MockNode()
        collector = MessageCollector(node, "/test_topic", MockMessage)

        assert len(node._subscriptions) == 1
        collector.destroy()
        assert len(node._subscriptions) == 0


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: test/cpp/test_message_collector.cpp
// Copyright 2026 John Vial
// SPDX-License-Identifier: Apache-2.0

#include <gtest/gtest.h>
#include <thread>

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include "sim_harness/core/message_collector.hpp"

using sim_harness::MessageCollector;

class MessageCollectorTest : public ::testing::Test
{
protected:
  void SetUp() override
  {
    node_ = std::make_shared<rclcpp::Node>("test_node");
    executor_.add_node(node_);
  }

  void TearDown() override
  {
    executor_.remove_node(node_);
    node_.reset();
  }

  rclcpp::Node::SharedPtr node_;
  rclcpp::executors::SingleThreadedExecutor executor_;
};

TEST_F(MessageCollectorTest, BasicCollection)
{
  auto collector = std::make_shared<MessageCollector<std_msgs::msg::String>>(
    node_, "/test_topic");

  // Create publisher
  auto pub = node_->create_publisher<std_msgs::msg::String>("/test_topic", 10);

  // Initially no messages
  EXPECT_EQ(collector->count(), 0u);
  EXPECT_FALSE(collector->hasMessages());

  // Publish a message
  std_msgs::msg::String msg;
  msg.data = "hello";
  pub->publish(msg);

  // Spin to receive
  for (int i = 0; i < 10 && collector->count() == 0; ++i) {
    executor_.spin_some(std::chrono::milliseconds(10));
  }

  EXPECT_EQ(collector->count(), 1u);
  EXPECT_TRUE(collector->hasMessages());

  auto messages = collector->getMessages();
  ASSERT_EQ(messages.size(), 1u);
  EXPECT_EQ(messages[0].data, "hello");
}

TEST_F(MessageCollectorTest, MultipleMessages)
{
  auto collector = std::make_shared<MessageCollector<std_msgs::msg::String>>(
    node_, "/test_topic");

  auto pub = node_->create_publisher<std_msgs::msg::String>("/test_topic", 10);

  // Publish multiple messages
  for (int i = 0; i < 5; ++i) {
    std_msgs::msg::String msg;
    msg.data = "message_" + std::to_string(i);
    pub->publish(msg);
    executor_.spin_some(std::chrono::milliseconds(10));
  }

  // Give time for all messages to arrive
  for (int i = 0; i < 20 && collector->count() < 5; ++i) {
    executor_.spin_some(std::chrono::milliseconds(10));
  }

  EXPECT_GE(collector->count(), 5u);
}

TEST_F(MessageCollectorTest, GetLatest)
{
  auto collector = std::make_shared<MessageCollector<std_msgs::msg::String>>(
    node_, "/test_topic");

  // No messages yet
  EXPECT_FALSE(collector->getLatest().has_value());

  auto pub = node_->create_publisher<std_msgs::msg::String>("/test_topic", 10);

  std_msgs::msg::String msg1, msg2;
  msg1.data = "first";
  msg2.data = "second";

  pub->publish(msg1);
  // Wait for first message
  for (int i = 0; i < 10 && collector->count() < 1; ++i) {
    executor_.spin_some(std::chrono::milliseconds(10));
  }

  pub->publish(msg2);
  // Wait for second message
  for (int i = 0; i < 10 && collector->count() < 2; ++i) {
    executor_.spin_some(std::chrono::milliseconds(10));
  }

  auto latest = collector->getLatest();
  ASSERT_TRUE(latest.has_value());
  EXPECT_EQ(latest->data, "second");
}

TEST_F(MessageCollectorTest, Clear)
{
  auto collector = std::make_shared<MessageCollector<std_msgs::msg::String>>(
    node_, "/test_topic");

  auto pub = node_->create_publisher<std_msgs::msg::String>("/test_topic", 10);

  std_msgs::msg::String msg;
  msg.data = "test";
  pub->publish(msg);

  for (int i = 0; i < 10 && collector->count() == 0; ++i) {
    executor_.spin_some(std::chrono::milliseconds(10));
  }

  EXPECT_GT(collector->count(), 0u);

  collector->clear();

  EXPECT_EQ(collector->count(), 0u);
  EXPECT_FALSE(collector->hasMessages());
}

TEST_F(MessageCollectorTest, GetTopic)
{
  auto collector = std::make_shared<MessageCollector<std_msgs::msg::String>>(
    node_, "/my/test/topic");

  EXPECT_EQ(collector->getTopic(), "/my/test/topic");
}

TEST_F(MessageCollectorTest, Callback)
{
  auto collector = std::make_shared<MessageCollector<std_msgs::msg::String>>(
    node_, "/test_topic");

  bool callback_called = false;
  std::string received_data;

  collector->setCallback([&](const std_msgs::msg::String & msg) {
      callback_called = true;
      received_data = msg.data;
    });

  auto pub = node_->create_publisher<std_msgs::msg::String>("/test_topic", 10);

  std_msgs::msg::String msg;
  msg.data = "callback_test";
  pub->publish(msg);

  for (int i = 0; i < 10 && !callback_called; ++i) {
    executor_.spin_some(std::chrono::milliseconds(10));
  }

  EXPECT_TRUE(callback_called);
  EXPECT_EQ(received_data, "callback_test");
}

int main(int argc, char ** argv)
{
  testing::InitGoogleTest(&argc, argv);
  rclcpp::init(argc, argv);
  int result = RUN_ALL_TESTS();
  rclcpp::shutdown();
  return result;
}
